biochatter/api_agent/blast.py

```diff
@@ -0,0 +1,352 @@
+            "DATABASE": request_data.database,
+            "QUERY": request_data.query,
+            "FORMAT_TYPE": request_data.format_type,
+            "MEGABLAST": request_data.megablast,
+            "HITLIST_SIZE": request_data.max_hits,
+        }
+        # Include any other_params if provided
+        if request_data.other_params:
+            data.update(request_data.other_params)
+        # Make the API call
+        query_string = urlencode(data)
+        # Combine base URL with the query string
+        full_url = f"{request_data.url}?{query_string}"
+        # Print the full URL
+        request_data.full_url = full_url
+        print("Full URL built by retriever:\n", request_data.full_url)
+        response = requests.post(request_data.url, data=data)
+        response.raise_for_status()
+        # Extract RID from response
+        print(response)
+        match = re.search(r"RID = (\w+)", response.text)
+        if match:
+            return match.group(1)
+        else:
+            raise ValueError("RID not found in BLAST submission response.")
+
+    def fetch_and_save_results(
+        self,
+        question_uuid: uuid,
+        query_return: str,
+        save_path: str,
+        max_attempts: int = 10000,
+    ):
+        """SECOND function to be called for a BLAST query.
+        Will look for the RID to fetch the data
+        """
+        file_name = f"BLAST_results_{question_uuid}.txt"
+        ###
+        ###    TO DO: Implement logging for all BLAST queries
+        ###
+        # log_question_uuid_json(request_data.question_uuid,question, file_name, save_path, log_file_path,request_data.full_url)
+        base_url = "https://blast.ncbi.nlm.nih.gov/Blast.cgi"
+        check_status_params = {
+            "CMD": "Get",
+            "FORMAT_OBJECT": "SearchInfo",
+            "RID": query_return,
+        }
+        get_results_params = {
+            "CMD": "Get",
+            "FORMAT_TYPE": "XML",
+            "RID": query_return,
+        }
+        # checking path: should end with '/'
+        if not save_path.endswith("/"):
+            save_path += "/"
+        # Check the status of the BLAST job
+        for attempt in range(max_attempts):
+            status_response = requests.get(base_url, params=check_status_params)
+            status_response.raise_for_status()
+            status_text = status_response.text
+            print("evaluating status")
+            if "Status=WAITING" in status_text:
+                print(f"{question_uuid} results not ready, waiting...")
+                time.sleep(15)
+            elif "Status=FAILED" in status_text:
+                with open(f"{save_path}{file_name}", "w") as file:
+                    file.write("BLAST query FAILED.")
+            elif "Status=UNKNOWN" in status_text:
+                with open(f"{save_path}{file_name}", "w") as file:
+                    file.write("BLAST query expired or does not exist.")
+                raise
```

Comment:
is this code intentional?

Reply:
This is a good point. For now i have it like this (i.e. the error is written in the blast result file) because it makes it easier to evaluate the results in downstream testing of the performance of the api agent: **did the query fail or the interpretation of the content of the BLAST file**. Optimally we would indeed handle this differently but it would require another logic. For now i would suggest to keep it as like this.

Reply:
This also made me confused, why do we need to write result to a file instead of returning it? This makes caller difficult to parse result, particularly for GptConversation, which have to read it and pass it to LLM.
Is there something I might have overlooked?

Reply:
I understand your confusion, the optimal way of handling this would indeed be to implement it as a return instead of writing to file. 
This is a 'legacy code' from my initial implementation where I decoupled the query+write return to file and answer extraction to speed up the testing (GeneTuring test:  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC10054955/). But bottom line: you are right and I will change as soon as i have time for it.